parser/options.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+type options struct {
+	maxRecursionDepth  int
+	errorRecoveryLimit int
+	macros             map[string]Macro
+}
+
+// Option configures the behavior of the parser.
+type Option func(*options)
+
+// WithMaxRecursionDepth is an option which limits the maximum depth the parser will attempt to
+// parse the expression before giving up.
+func WithMaxRecursionDepth(maxRecursionDepth int) Option {
+	return func(opts *options) {
+		opts.maxRecursionDepth = maxRecursionDepth
+	}
+}
+
+// WithErrorRecoveryLimit is an option which limits the number of attempts the parser will perform
+// to recover from an error.
+func WithErrorRecoveryLimit(errorRecoveryLimit int) Option {
+	return func(opts *options) {
+		opts.errorRecoveryLimit = errorRecoveryLimit
+	}
+}
+
+// WithMacro adds the given macro to the parser.
+func WithMacro(macro Macro) Option {
+	return func(opts *options) {
+		if macro != nil {
+			if opts.macros == nil {
+				opts.macros = make(map[string]Macro)
+			}
+			opts.macros[macro.MacroKey()] = macro
+		}
+	}
+}
+
+// WithMacros adds the given macros to the parser.
+func WithMacros(macros ...Macro) Option {
+	return func(opts *options) {
+		for _, m := range macros {
+			if m != nil {
+				if opts.macros == nil {
+					opts.macros = make(map[string]Macro)
+				}
+				opts.macros[m.MacroKey()] = m
+			}
+		}
+	}
+}
+
+func applyOptions(ropts *options, opts ...Option) {
+	for _, opt := range opts {
```

Comment:
I would just do this in place in the `newParser` call.

Reply:
Done.